crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1381,6 +1429,12 @@ impl<'db> Type<'db> {
                 // are both of type Literal[345], for example.
                 false
             }
+
+            // A typevar is never a singleton, since it can always be specialized to `Any`.
+            // QUESTION: Should a bounded typevar be considered a singleton if its bound is? It can
+            // still be specialized to `Any`, instead of to the singleton bound.
```

Comment:
Open question

Reply:
> Should a bounded typevar be considered a singleton if its bound is?

I don't think so? Even without gradual types, you could always assign `Never`, which is also not a singleton.

Reply:
> For _constrained_ typevars, we might be able to return `true` in some cases (if all possible options are singleton types)?

Yeah I wasn't sure how "viral" to make the constraints in this case. You can still specialize the typevar as `Any`, which is not in general a singleton. But do we want to carry through the knowledge that that `Any` can really only be materialized to one of the constraint types?

And if we do, do we want to that here too?:

```py
def f[T: (int, str)](t: T):
    reveal_type(t.__class__) # revealed: Literal[int | str]
    a: Any = t
    reveal_type(a.__class__) # revealed: Any or Literal[int | str]?
```

(Though I could see the argument that the explicit `Any` annotation is a request to ignore the constraints for some reason?)

Reply:
I've implemented the interpretation that a constrained typevar is singleton/single-valued if all of the constraint types are.  I think that is not really the same as my "assign to an explicitly `Any`-annotated variable" case.